RFC/src/RFC-0250_Covenants.md

```diff
@@ -0,0 +1,468 @@
+Covenents aims to provide restrictions on the _future_ spending of subsequent transactions to enable a number of powerful 
+use-cases, such as
+- [vaults]
+- side-chain checkpointing transactions,
+- commission on NFT transfers, and 
+- many others not thought of here.
+
+## Related Requests for Comment
+
+- [RFC-0200: Base Layer Extensions](BaseLayerExtensions.md)
+- [RFC-0300: The Tari Digital Assets Network](RFC-0300_DAN.md)
+
+## Introduction
+
+The Tari protocol already provides programmable consensus, through [TariScript], that restricts whether a [UTXO]
+may be included as an input to a transaction (a.k.a spent). The scope of information [TariScript] is inherently limited,
+by the [TariScript Opcodes] and the input data provided by a spender. Once the requirements of the script are met, 
+a spender may generate [UTXO]s of their choosing, within the constraints of [MimbleWimble]. 
+
+This RFC aims to expand the capabilities of Tari protocol by adding _additional requirements_, called covenants 
+that allow the owner(s) of a [UTXO] to control the composition of a _subsequent_ transaction. 
+
+Covenants are not a new idea and have been proposed and implemented in various forms by others.
+
+For example,
+- [Bitcoin-NG covenents] put forward the `CheckOutputVerify` script opcode.
+- [Handshake] has implemented covenants to add  the [UTXO] state of their auctioning process.
+- [Elements Covenants]
+
+## Covenants in MimbleWimble 
+
+In block chains like Bitcoin, a block contains discrete transactions containing inputs and outputs. A covenant 
+in Bitcoin would be able to interrogate those outputs _belonging to the input_ to ensure that they adhere to rules.
+
+In [MimbleWimble], the body of a block and transaction can be expressed in an identical data structure. This
+is indeed the case in the [Tari codebase], which defines a structure called `AggregateBody` containing inputs
+and outputs (and kernels) for transactions and blocks. This is innate to [MimbleWimble], so even if we were 
+to put a "box" around these inputs/outputs there is nothing to stop someone from including inputs and 
+outputs from other boxes as long as balance is maintained.
+
+This results in an interesting dilemma: how do we allow rules that dictate how future outputs look only armed with
+the knowledge that the rule must apply to one or more outputs?
+
+In this RFC, we propose a covenant scheme that allows the [UTXO] originator to express a _filter_ that must be 
+satisfied for a subsequent spending transaction to be considered valid.
+
+## Assumptions
+
+The following assumptions are made:
+1. Duplicate commitments within a block are disallowed by consensus _prior_ to covenant execution,
+2. all outputs in the output set are valid, and
+3. all inputs are valid spends, save for covenent checks.
+
+## Protocol modifications
+
+Modifications to the existing protocol and consensus are as follows:
+
+- the covenant is recorded in the transaction [UTXO],
```

Comment:
There is not really a thing such as a transaction, there are inputs and outputs, and kernels that are grouped together + some header info. 
Is this recorder, in the input, output, kernel, or the header (such show the offset and script offset is stored)